lib/bolt/plugin/puppetdb.rb

```diff
@@ -16,8 +16,40 @@ def hooks
       end
 
       def lookup_targets(opts)
-        nodes = @puppetdb_client.query_certnames(opts['query'])
-        nodes.map { |certname| { 'uri' => certname } }
+        targets = @puppetdb_client.query_certnames(opts['query'])
+        facts = []
+
+        if opts['config']
+          # This is done in parts instead of in place so that we only need to
+          # make one puppetDB query
+          Bolt::Util.walk_vals(opts['config']) do |value|
+            if value.is_a?(String)
+              facts << value.split(".")
+            end
+            value
+          end
+          fact_values = @puppetdb_client.fact_values(targets, facts)
+        end
+
+        targets.map do |certname|
+          target = { 'uri' => certname }
+          if opts['config']
```

Comment:
The ticket description is kind of confusing. The title `Allow puppetdb plugin to set host and name independently` and first paragraph of description
```
The puppetdb plugin will only accept a query that returns a certname and then set that certname to be the targets uri. For some users the node is not accessible through the certname so they need to lookup an IP address and set the name and certname of the target separately.
```
Make is seem like you should be able to set `name` or `uri` in addition to `config` like in the terraform plugin. This condition looks like you can only set config values from facts.